chore: publish docs to website

lidavidm opened the pull request.

lidavidm commented:
CC @nealrichardson I know you wanted to have `/docs/adbc` and `/docs/dev/adbc`, but it looks like .asf.yaml can only publish to a single subdirectory. So I've reverted to publishing to `/adbc` - maybe we can add redirects in the future? (Right now it'll end up as [`/adbc/dev` for the dev docs](https://github.com/apache/arrow-adbc/tree/asf-site), and nothing else, with my plan being to add `/adbc/0.1.0` etc. for each version.)

lidavidm commented:
CC @kou does this look OK? Maybe this should instead go under `/adbc/docs` so that in apache/arrow-site, we can generate an `/adbc/index.html` in Jekyll…

lidavidm commented:
Updated.

kou commented:
Why do we want to generate `/adbc/index.html` in Jekyll?
I think that it's better that we generate all files for https://arrow.apache.org/adbc/ in apache/arrow-adbc.

How about putting the latest documents to https://arrow.apache.org/adbc/ and the dev documents to https://arrow.apache.org/adbc/dev/ like https://arrow.apache.org/docs/ and https://arrow.apache.org/docs/dev/ ?

BTW, we need to copy `.asf.yaml` in `main` to `asf-site` to deploy https://arrow.apache.org/adbc/ .
See also: https://github.com/apache/arrow-site/pull/115 and https://github.com/apache/arrow-site/pull/116

lidavidm commented:
> Why do we want to generate `/adbc/index.html` in Jekyll? I think that it's better that we generate all files for https://arrow.apache.org/adbc/ in apache/arrow-adbc.

I was thinking about having some sort of (brief) landing page, but it is redundant. (Or maybe a redirect, or a version selector.)

> How about putting the latest documents to https://arrow.apache.org/adbc/ and the dev documents to https://arrow.apache.org/adbc/dev/ like https://arrow.apache.org/docs/ and https://arrow.apache.org/docs/dev/ ?

I wanted to have versioned documentation from the start, so it seems weird to me to have them nested like that. (I still need to look at different approaches to that, but something like https://github.com/steinwurf/versjon assumes each version falls in its own subdirectory.) But consistency probably matters more.

> BTW, we need to copy `.asf.yaml` in `main` to `asf-site` to deploy https://arrow.apache.org/adbc/ . See also: [apache/arrow-site#115](https://github.com/apache/arrow-site/pull/115) and [apache/arrow-site#116](https://github.com/apache/arrow-site/pull/116)

Ah, thank you.

kou commented:
> I was thinking about having some sort of (brief) landing page, but it is redundant. (Or maybe a redirect, or a version selector.)

Ah, it makes sense.
I used the "redirect" approach in other project: https://hayamizu-lab.github.io/treefit-python/ (source: https://github.com/hayamizu-lab/treefit-python/blob/master/docs/index.html )

> I wanted to have versioned documentation from the start, so it seems weird to me to have them nested like that.

I agree with you. FYI: I like the following URLs:

* https://apache.arrow.org/adbc/latest/ : The latest release (This is a copy of https://apache.arrow.org/adbc/X.Y.Z/ . Users can use both of https://apache.arrow.org/adbc/latest/ and https://apache.arrow.org/adbc/1.0.0/ when the latest release is 1.0.0)
* https://apache.arrow.org/adbc/dev/ or https://apache.arrow.org/adbc/main/ : For the `main` branch
* https://apache.arrow.org/adbc/1.0.0/ : For 1.0.0

kou commented:
We can use apache/arrow-adbc for the "redirect" approach because we just need to put a simple HTML. We don't need to use the same design in https://arrow.apache.org/ for the approach.

lidavidm commented:
Cool. I've reverted it to deploy to adbc/. Also, I added https://github.com/steinwurf/versjon which generates the index.html with a redirect for us. It also embeds a version switcher into the docs, and generates a `stable/index.html` with a redirect (if it detects a stable version).

lidavidm commented:
Ah, versjon is not idempotent, hmm.

lidavidm commented:
I'll take a look at embedding it ourselves

lidavidm commented:
Ok…added a homegrown version switcher. It embeds a reference to a JavaScript file in the generated docs. The JavaScript file will sit in `adbc/version.js` and will have to be manually added to the asf-site branch. The contents are this:

<details>

```javascript
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

window.addEventListener("DOMContentLoaded", () => {
    // Injected by template
    window.fetch(versionListingPath)
        .then((r) => r.text())
        .then((text) => {
            const root = document.querySelector("#version-switcher ul");
            const versions = text
                  .trim()
                  .split(/\n/g)
                  .map((version) => {
                      return version.split(/;/);
                  });
            const versionRegex = /^([0-9]+)\.([0-9]+)\.([0-9]+)(.*)$/;
            versions.sort((x, y) => {
                const lhs = x[1].match(versionRegex);
                const rhs = y[1].match(versionRegex);
                if (lhs === null && rhs === null) {
                    return x[1].localeCompare(y[1]);
                }
                if (lhs === null) return 1;
                if (rhs === null) return -1;
                // Major
                const lhsMajor = parseInt(lhs[1], 10);
                const rhsMajor = parseInt(rhs[1], 10);
                if (lhsMajor < rhsMajor) {
                    return -1;
                } else if (lhsMajor > rhsMajor) {
                    return 1;
                }

                const lhsMinor = parseInt(lhs[2], 10);
                const rhsMinor = parseInt(rhs[2], 10);
                if (lhsMinor < rhsMinor) {
                    return -1;
                } else if (lhsMinor > rhsMinor) {
                    return 1;
                }

                const lhsPatch = parseInt(lhs[3], 10);
                const rhsPatch = parseInt(rhs[3], 10);
                if (lhsPatch < rhsPatch) {
                    return -1;
                } else if (lhsPatch > rhsPatch) {
                    return 1;
                }

                return lhs[4].localeCompare(rhs[4]);
            });
            versions.forEach((version) => {
                const el = document.createElement("a");
                el.setAttribute("href", versionsRoot + "/" + version[0]);
                el.innerText = version[1];
                if (version[1] === currentVersion) {
                    el.classList.toggle("active");
                }
                const li = document.createElement("li");
                li.appendChild(el);
                root.appendChild(li);
            });
        });
});
```

</details>

It just populates a list embedded in the docs. The generated docs style the list. The versions are pulled from a text file that we can generate in CI.

lidavidm commented:
It looks like this:

![image](https://user-images.githubusercontent.com/327919/207670484-7cca6cc4-e009-434c-9c5f-3323eae1066c.png)

![image](https://user-images.githubusercontent.com/327919/207670637-8a8a6e92-d5e1-4376-ba78-aa62121642e0.png)

(screenshot doesn't capture it, but it scrolls)

lidavidm commented:
Unlike versjon, because it doesn't modify the docs HTML, it is idempotent. We can update the switcher in all docs just by updating the text file and/or JavaScript file. (We could save a round-trip with some more acrobatics, by templating the JavaScript file with the contents of the text file, but we can always do that later without having to regenerate docs.)

lidavidm commented:
Demo here: https://github.com/lidavidm/arrow-adbc/tree/asf-site

(check out the branch and run `python -m http.server 12345`, then visit localhost:12345)

And then in the release PR, I can have the action generate an `index.html` and a `latest` that redirect to the docs from the newly pushed tag. (I suppose that'll backfire if we eventually do patch releases/we'll eventually need a version comparator there.)

lidavidm commented:
Also updated that branch with the results of pushing a tag (which also generated index.html and latest/)

kou reviewed (approved):
+1

kou commented:
> The JavaScript file will sit in `adbc/version.js` and will have to be manually added to the asf-site branch.

How about maintaining it in the main branch and just copying it to the asf-site branch on publish?

lidavidm commented:
Ok! In that case I'll also move this Bash code out into actual scripts

lidavidm commented:
Updated!

1. Bash scripts in YAML replaced with an actual script
2. Version list embedded in the JS file to avoid a network round-trip
3. Version list pre-sorted in Bash so we don't have to duplicate it in JavaScript
4. Script handles tagged versions so we can avoid duplicating the action code for release

lidavidm commented:
Demo here: https://github.com/lidavidm/arrow-adbc/tree/asf-site

Testing out the tag workflow here: https://github.com/lidavidm/arrow-adbc/actions/runs/3705059422

lidavidm commented:
Hmm, one bug in the script now that I need to fix...

lidavidm commented:
And https://github.com/apache/arrow-adbc/pull/246 to prepare for deployment

kou reviewed (approved):
+1

lidavidm commented:
Thanks so much for your help!

lidavidm commented:
It works! https://arrow.apache.org/adbc/
